scipy/integrate/_ivp/radau.py

```diff
@@ -467,14 +503,39 @@ def _step_impl(self):
             ZE = Z.T.dot(E) / h
             error = self.solve_lu(LU_real, f + ZE)
             scale = atol + np.maximum(np.abs(y), np.abs(y_new)) * rtol
-            error_norm = norm(error / scale)
+            if self.mass_matrix is None:
+                error = self.solve_lu(LU_real, f + ZE)
+                error_norm = norm(error / scale)
+            else:  # see [1], chapter IV.8, page 127
+                error = self.solve_lu(LU_real, f + self.mass_matrix.dot(ZE))
+                if self.index_algebraic_vars is not None:
+                    # correct for the overestimation of the error on
+                    # algebraic variables, ideally multiply their errors by
+                    # (h ** index)
+                    error[self.index_algebraic_vars] = 0.
+                    error_norm = np.linalg.norm(error / scale) / (n - self.nvars_algebraic) ** 0.5
+                    # we exclude the algebraic components, otherwise
+                    # they artificially lower the error norm
+                else:
+                    error_norm = norm(error / scale)
+
             safety = 0.9 * (2 * NEWTON_MAXITER + 1) / (2 * NEWTON_MAXITER
                                                        + n_iter)
 
             if rejected and error_norm > 1:
-                error = self.solve_lu(LU_real, self.fun(t, y + error) + ZE)
-                error_norm = norm(error / scale)
-
+                if self.mass_matrix is None:
```

Comment:
This looks pretty similar to the block above. Consider whether it's worth it to write a private function to reduce code duplication.

Reply:
These are actually two different error estimates. They were originally present in the code, and are quite necessary for a proper functioning of the error estimate. Basically, the second one (which involves on linear system resolution, unlike the first one which is explicit) is less affected by the stiffness of the system. It is used only when the first estimate is too high. This is also done in the original Fortran code from Hairer & Wanner. I have added a comment to clarify that.